peps/pep-0747.rst

```diff
@@ -711,12 +735,38 @@ assigned to variables and manipulated like any other data in a program:
 ``TypeExpr[]`` is how you spell the type of a variable containing a
 type annotation object describing a type.
 
-``TypeExpr[]`` is similar to ``type[]``, but ``type[]`` can only used to
+``TypeExpr[]`` is similar to ``type[]``, but ``type[]`` can only
```

Comment:
I understand the point you're trying to make here, but it's a little misleading because `type` (when used in a type expression) works with many of the example in this list including `type[list[int]]` or `type[int | None]`.

Reply:
Huh. Indeed I see the following passes typechecking in both mypy and pyright:

```
t1: type[list]
t2: type[list[int]]
t3: type[int | None]
```

It's quite surprising to me that lines `t2` and `t3` are accepted. PEP 484 says "The value corresponding to Type[C] must be an actual class object that’s a subtype of C, not a special form." I'm pretty sure neither `list[int]` nor `int | None` are class objects. They're both special forms: GenericAlias and UnionType respectively.

So maybe `type[C]` actually accepts some set of objects including all class objects but not including all forms spellable by a type expression? Very confusing. The current version of the typing specification does not appear to describe which `C` expressions are legal in `type[C]`.

Reply:
I'm inclined to think that both the `t2` and `t3` *should* be marked as an error and the current behavior of mypy / pyright accepting them is a bug. At least it's not a behavior that a program should depend on, based on what PEP 484 says.

Similar bugs were opened against mypy where `type` was improperly matching various kinds of non-class objects, prompting the creation of this PEP in the first place.

Reply:
> They're both special forms: GenericAlias and UnionType respectively.

These are not special forms. `GenericAlias` and `UnionTypes` are classes that are used internally at runtime to implement typing constructs. Special forms are documented typing symbols that are valid within type expressions. They're generally defined in the `typing` module and are represented in the typeshed stub as `_SpecialForm`. Type checkers include custom logic to handle these special forms. By contrast, type checkers have no knowledge of `GenericAlias`, and they have minimal knowledge of (and would preferably have no knowledge of) `UnionType`.

From the perspective of a static type checker, the the type expression `list[int]` is a class, and it's compatible with `type`. It is instantiable via a call and is a subtype of `type`. (At runtime, it happens to be represented by a `GenericAlias`, but that's an implementation detail.)